[nnc] Move device type checks from isSupported to typesAreSupported

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55213 Integrate NNC conv2d with fuser
* **#55025 [nnc] Move device type checks from isSupported to typesAreSupported**
* #54920 NNC depthwise conv2d implementation

Something needs to be fixed about the names of these functions,
because they are confusing.

The profiling infrastructure calls `isSupported` to see if it should insert
profiling nodes.

The fuser calls `isSupported` but also `typesAreSupported` to determine if it
can actually fuse the node.

At profiling time, we don't know device types yet, so we can't use device type
checks in `isSupported` or else we'll never profile the node.  So we want to
move those checks into `typesAreSupported`, where we actually have profiling
info available.

Differential Revision: [D27454968](https://our.internmc.facebook.com/intern/diff/D27454968/)

bertmaher committed:
[nnc] Move device type checks from isSupported to typesAreSupported

Something needs to be fixed about the names of these functions,
because they are confusing.

The profiling infrastructure calls `isSupported` to see if it should insert
profiling nodes.

The fuser calls `isSupported` but also `typesAreSupported` to determine if it
can actually fuse the node.

At profiling time, we don't know device types yet, so we can't use device type
checks in `isSupported` or else we'll never profile the node.  So we want to
move those checks into `typesAreSupported`, where we actually have profiling
info available.

Differential Revision: [D27454968](https://our.internmc.facebook.com/intern/diff/D27454968/)

[ghstack-poisoned]

eellison reviewed (commented):
For conv2d, we're only supporting depthwise convs right? Dont we need to add more checks so that we're not fusing conv2d more generally ?

bertmaher committed:
fix format on "[nnc] Move device type checks from isSupported to typesAreSupported"


Something needs to be fixed about the names of these functions,
because they are confusing.

The profiling infrastructure calls `isSupported` to see if it should insert
profiling nodes.

The fuser calls `isSupported` but also `typesAreSupported` to determine if it
can actually fuse the node.

At profiling time, we don't know device types yet, so we can't use device type
checks in `isSupported` or else we'll never profile the node.  So we want to
move those checks into `typesAreSupported`, where we actually have profiling
info available.

Differential Revision: [D27454968](https://our.internmc.facebook.com/intern/diff/D27454968/)

[ghstack-poisoned]

ZolotukhinM commented:
> For conv2d, we're only supporting depthwise convs right? Dont we need to add more checks so that we're not fusing conv2d more generally ?

@eellison As of now, we don't fuse convolutions:
https://github.com/pytorch/pytorch/blob/a0ae3e520fa36a38a081ff08d102f3f569a7dc8a/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L164-L166

When we decide to enable it, we need to add appropriate checks, yes.

bertmaher committed:
TORCH_API on "[nnc] Move device type checks from isSupported to typesAreSupported"


Something needs to be fixed about the names of these functions,
because they are confusing.

The profiling infrastructure calls `isSupported` to see if it should insert
profiling nodes.

The fuser calls `isSupported` but also `typesAreSupported` to determine if it
can actually fuse the node.

At profiling time, we don't know device types yet, so we can't use device type
checks in `isSupported` or else we'll never profile the node.  So we want to
move those checks into `typesAreSupported`, where we actually have profiling
info available.

Differential Revision: [D27454968](https://our.internmc.facebook.com/intern/diff/D27454968/)

[ghstack-poisoned]

bertmaher commented:
> For conv2d, we're only supporting depthwise convs right? Dont we need to add more checks so that we're not fusing conv2d more generally ?

Yeah, it's commented-out in the list of supported ops above so we won't fuse it (until the next diff in this stack :-) ).

bertmaher committed:
rebase on "[nnc] Move device type checks from isSupported to typesAreSupported"


Something needs to be fixed about the names of these functions,
because they are confusing.

The profiling infrastructure calls `isSupported` to see if it should insert
profiling nodes.

The fuser calls `isSupported` but also `typesAreSupported` to determine if it
can actually fuse the node.

At profiling time, we don't know device types yet, so we can't use device type
checks in `isSupported` or else we'll never profile the node.  So we want to
move those checks into `typesAreSupported`, where we actually have profiling
info available.

Differential Revision: [D27454968](https://our.internmc.facebook.com/intern/diff/D27454968/)

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -934,6 +913,28 @@ class TensorExprFuser {
         }
 
+        // Operator is only supported on CUDA.
+        if (node->isMemberOf(cuda_only_operator_set)) {
```
could we move this outside of `for`, otherwise we seem to be re-checking for every value?

Krovatkin replied:
we omitted checking the output types which is probably okay since these ops aren't c-tor ones.

Krovatkin reviewed (approved):
:shipit:

bertmaher committed:
rebase on "[nnc] Move device type checks from isSupported to typesAreSupported"


Something needs to be fixed about the names of these functions,
because they are confusing.

The profiling infrastructure calls `isSupported` to see if it should insert
profiling nodes.

The fuser calls `isSupported` but also `typesAreSupported` to determine if it
can actually fuse the node.

At profiling time, we don't know device types yet, so we can't use device type
checks in `isSupported` or else we'll never profile the node.  So we want to
move those checks into `typesAreSupported`, where we actually have profiling
info available.

Differential Revision: [D27454968](https://our.internmc.facebook.com/intern/diff/D27454968/)

[ghstack-poisoned]
